# Final Project Machine Learning: Machine Learning for the Image Classification using Tensorflow

- Name: Stefanus Bernard Melkisedek
- Email: stefanussipahutar@gmail.com
- ID Dicoding: stefansphtr


**Note:** This notebook is meant to be run in VS Code. If you're using a different environment, you may need to adjust some of the code.


In [16]:
# GUI related modules
import tkinter as tk
from tkinter import filedialog

# Machine learning and data processing modules
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.preprocessing import image
from keras.preprocessing.image import load_img
from sklearn.model_selection import train_test_split
import numpy as np

# Image and graph plotting modules
import matplotlib.pyplot as plt
import matplotlib.image as mpimg

# File and OS related modules
import zipfile
import os
import shutil
import random

In [2]:
# Define the path to the ZIP file
local_zip = "../../data/rockpaperscissors.zip"

# Open the ZIP file in read mode
zip_ref = zipfile.ZipFile(local_zip, "r")

# Extract all files from the ZIP file to the /tmp dir
zip_ref.extractall("/tmp")

# Close the ZIP file
zip_ref.close()

# Define the base directory where the images are stored
base_dir = "/tmp/rockpaperscissors/rps-cv-images"

# Define directories
rock_dir = os.path.join(base_dir, "rock")
paper_dir = os.path.join(base_dir, "paper")
scissors_dir = os.path.join(base_dir, "scissors")

In [3]:
# Get the list of all files and directories in the `rock_dir` directory
os.listdir(rock_dir)

['00nKV8oHuTGi20gq.png',
 '01dM3ewPIsnBICu1.png',
 '02vG75hQW9Vp4oTl.png',
 '0bioBZYFCXqJIulm.png',
 '0NDYNEoDui7o64gU.png',
 '0OEXfEooCXlljaEl.png',
 '0P6uxM8Vr1DwySHe.png',
 '19H63c0MWD56EWNq.png',
 '1BqjHe5igJAgUjiN.png',
 '1COBZEE1ALiJIivQ.png',
 '1DxbXT3M2qiMjCmC.png',
 '1e1VCnCEApnFh9Oo.png',
 '1mnAKQCHEDbtjPX2.png',
 '1MOm36DHK0R8OfIC.png',
 '1UucZqS3qblvU0cH.png',
 '1UXuUYJ4rVj2eSvt.png',
 '1VoYiUgPE6h45MLu.png',
 '1vvcitV1s17gKdbn.png',
 '1wk2Hl1Ih9guOwx6.png',
 '1WuMUpzupKs67q29.png',
 '2affjOmZChc9AXpR.png',
 '2DdERUV7Te8ivywq.png',
 '2EAkMCm7ZLnNoBQ7.png',
 '2f25VA1nfxiC86xj.png',
 '2HSPQEteONUjEXAF.png',
 '2j9iOcnxJzcyaFB5.png',
 '2JRYQXb5kmlhgD5i.png',
 '2MKZD0VUw7Tez2Jx.png',
 '2NmrcDGkc7FQuu12.png',
 '2nSUlcvPf1DOCo1j.png',
 '2O9XPBJRT119drWX.png',
 '2Pt5UNQkMzXLtbEp.png',
 '2UaJz6Z3xJv3WX9h.png',
 '2uVMdMj10yjpR6Py.png',
 '2uxDLQKemRqulhIx.png',
 '2vwLOZSd8FCXbuck.png',
 '32PDtFdAkUiAJbmP.png',
 '337ARHTZmhCSkoEM.png',
 '3aRXheNUQ6cxechp.png',
 '3k7WrSAInCaatFtl.png',


In [4]:
# Get the list of all files and directories in the `paper_dir` directory
os.listdir(paper_dir)

['04l5I8TqdzF9WDMJ.png',
 '0a3UtNzl5Ll3sq8K.png',
 '0cb6cVL8pkfi4wF6.png',
 '0eqArS2GgsBeqgSn.png',
 '0Og76sl5CJhbxWWx.png',
 '0RA9rcrv8iVvuDDU.png',
 '0t08v9bAaA0mXGCQ.png',
 '0Uomd0HvOB33m47I.png',
 '0vugygEjxQJPr9yz.png',
 '0zKU6wEhT2cDe9j0.png',
 '14K62H4KG1hg5qiJ.png',
 '1Dfw5FRlS4h60JPX.png',
 '1fKclJk4eBWOQSR1.png',
 '1Frbe8cdOdkciOBg.png',
 '1jHtb93pI8f6pfKo.png',
 '1oKnUgV2CdCaQUUX.png',
 '1Tt9U9SzrOm98Xtw.png',
 '1uoW7QcH2qAgCTxv.png',
 '1yeoLbmp4alVOtFv.png',
 '1yKjzquSvl9ShK7K.png',
 '27XFtQjjXQ8AP8Tl.png',
 '2DQbLQfHlfrcATqi.png',
 '2F8Ng7620ANA7tEK.png',
 '2IOsxsG8AaxntdJM.png',
 '2OCAZGNv2OEzgAwU.png',
 '2PAcPusQ59xIMfiw.png',
 '2PiUTczN5Ez4VreF.png',
 '2SldJqjiZwIVPef9.png',
 '2tDgPZGqbrw95j39.png',
 '2UKnsncgPodQLOFR.png',
 '37PRV3B9HxuUPkQr.png',
 '389udM70tLrMVL5H.png',
 '3JN7OPsnxDBpcVeD.png',
 '3K6a8bG8j1YTH1l0.png',
 '3MMu7EXUtM6aEtXU.png',
 '3sKeYzYsjbWQVhc4.png',
 '3tNHP8R9YHru011u.png',
 '3tyVmhojjwETaf1u.png',
 '3XEQ73XMn7zaww2T.png',
 '3YvRoN0ZGuXVWEbh.png',


In [5]:
# Get the list of all files and directories in the `scissors_dir` directory/
os.listdir(scissors_dir)

['0657zSfiTYzP2jrl.png',
 '0CSaM2vL2cWX6Cay.png',
 '0ePX1wuCc3et7leL.png',
 '0Flw60Z2MAWWKn6S.png',
 '0Ug54ifXRqqlZS2Z.png',
 '0zoQAmDFXehOZsAp.png',
 '138Tx9KlEfheT3uB.png',
 '17HZDUFSVPxcar99.png',
 '1CXgK9fgGdSRggD9.png',
 '1i1dlQrE6JnhYXE4.png',
 '1Io5Ksm3bqy87EAU.png',
 '1jKhi65BPTLXnUI6.png',
 '1L1n6plr7jlZGirw.png',
 '1lEpWTJDphkm3HdC.png',
 '1MMkSW3f1CAsw03q.png',
 '1UojLz4CrT2G1Eo5.png',
 '1vQCX4af6hQmuwxZ.png',
 '1VRzspyXpQ6A2rKy.png',
 '1WX9KKoq0nkWiTxI.png',
 '1wZUIsowmZRmESjh.png',
 '1xkTkvyzmavV7lQ5.png',
 '1yDUM1w2HTbuPIfO.png',
 '256MrhkDete6aQhP.png',
 '277q9TUTAsBAbTCj.png',
 '2C9FEbBklwcVhF3W.png',
 '2DEYFsJ27UgVqv3W.png',
 '2E8j595xseKRF60J.png',
 '2EeIeEe1P0a0Fi9v.png',
 '2fxAdPTgrVIoITsL.png',
 '2gd0aqAxW55bHZP9.png',
 '2J7q0JbqJrjmZC05.png',
 '2l1K148aIJHRR1q7.png',
 '2M8LvUBGMOH1bsaz.png',
 '2TAGoXw7yaK0bXBu.png',
 '2tRxoWcFfrvjsHTZ.png',
 '2ufDQYIqKG1xRG7y.png',
 '2V5E7uHmAh9eK0qt.png',
 '2vDaPrc35RGC8nvM.png',
 '2ZPcSpOm7SEfQYwg.png',
 '3bkC6JAPog7xX9WO.png',


In [6]:
# Define a function to count the number of images in a directory
def count_images(directory):
    return len(os.listdir(directory))


# Count the number of images in each directory
num_rock_images = count_images(rock_dir)
num_paper_images = count_images(paper_dir)
num_scissors_images = count_images(scissors_dir)

# Show the results
print(f"Number of rock images: {num_rock_images}")
print(f"Number of paper images: {num_paper_images}")
print(f"Number of scissors images: {num_scissors_images}")

Number of rock images: 726
Number of paper images: 712
Number of scissors images: 750


In [7]:
# Define the directories for the training and validation datasets
data_train = os.path.join(base_dir, "training")
data_val = os.path.join(base_dir, "validation")

# Create the directories if they don't exist
os.makedirs(data_train, exist_ok=True)
os.makedirs(data_val, exist_ok=True)

# Define the proportion of data to be used for validation
validation_data_proportion = 0.4

# Split the data into training and validation sets for each category
train_rock_dir, val_rock_dir = train_test_split(
    os.listdir(rock_dir), test_size=validation_data_proportion
)
train_paper_dir, val_paper_dir = train_test_split(
    os.listdir(paper_dir), test_size=validation_data_proportion
)
train_scissors_dir, val_scissors_dir = train_test_split(
    os.listdir(scissors_dir), test_size=validation_data_proportion
)

In [8]:
# Function to move files from source directory to destination directory
def move_files_to_directory(files, source_directory, destination_directory):
    try:
        # Ensure the destination directory exists
        os.makedirs(destination_directory, exist_ok=True)

        # Move each file to the destination directory
        for file in files:
            shutil.move(
                os.path.join(source_directory, file),
                os.path.join(destination_directory, file),
            )
    except PermissionError:
        print(
            f"Permission denied for directory {destination_directory}. Please check the directory permissions."
        )
    except FileNotFoundError:
        print(
            f"Source or destination directory not found. Please check the directories."
        )
    except Exception as e:
        print(f"An error occurred: {e}")


# Define directories for each category
move_files_to_directory(train_rock_dir, rock_dir, os.path.join(data_train, "rock"))
move_files_to_directory(val_rock_dir, rock_dir, os.path.join(data_val, "rock"))

move_files_to_directory(train_paper_dir, paper_dir, os.path.join(data_train, "paper"))
move_files_to_directory(val_paper_dir, paper_dir, os.path.join(data_val, "paper"))

move_files_to_directory(
    train_scissors_dir, scissors_dir, os.path.join(data_train, "scissors")
)
move_files_to_directory(
    val_scissors_dir, scissors_dir, os.path.join(data_val, "scissors")
)

In [9]:
# Constants for ImageDataGenerator
RESCALE = 1.0 / 255
ROTATION_RANGE = 20
HORIZONTAL_FLIP = True
SHEAR_RANGE = 0.2
FILL_MODE = "nearest"

# Constants for flow_from_directory
TRAIN_DIR = data_train  # should contain 3 subdirectories, one for each class
TARGET_SIZE = (150, 150)
BATCH_SIZE = 20
CLASS_MODE = "categorical"  # 'categorical' for multi-class labels

# Initialize ImageDataGenerator
train_datagen = ImageDataGenerator(
    rescale=RESCALE,
    rotation_range=ROTATION_RANGE,
    horizontal_flip=HORIZONTAL_FLIP,
    shear_range=SHEAR_RANGE,
    fill_mode=FILL_MODE,
)

# Generate training data
train_generator = train_datagen.flow_from_directory(
    TRAIN_DIR, target_size=TARGET_SIZE, batch_size=BATCH_SIZE, class_mode=CLASS_MODE
)

Found 2143 images belonging to 3 classes.


In [10]:
# Constants for ImageDataGenerator
RESCALE = 1.0 / 255

# Constants for flow_from_directory
VAL_DIR = data_val  # should contain 3 subdirectories, one for each class
TARGET_SIZE = (150, 150)
BATCH_SIZE = 20
CLASS_MODE = "categorical"  # 'categorical' for multi-class labels

# Initialize ImageDataGenerator
test_datagen = ImageDataGenerator(rescale=RESCALE)

# Generate validation data
validation_generator = test_datagen.flow_from_directory(
    VAL_DIR, target_size=TARGET_SIZE, batch_size=BATCH_SIZE, class_mode=CLASS_MODE
)

Found 1938 images belonging to 3 classes.


In [11]:
# Create a dictionary to map the labels from the training generator
labels = {value: key for key, value in train_generator.class_indices.items()}

print(
    "Label Mappings for classes that are present in both the validation and training datasets:\n"
)
# Iterate over the labels dictionary and print each key-value pair
for key, value in labels.items():
    print(f"{key} : {value}")

Label Mappings for classes that are present in both the validation and training datasets:

0 : paper
1 : rock
2 : scissors


In [12]:
# Constants for the model
INPUT_SHAPE = (150, 150, 3)
FIRST_LAYER_FILTERS = 32
SECOND_LAYER_FILTERS = 64
THIRD_LAYER_FILTERS = 128
FOURTH_LAYER_FILTERS = 256
DENSE_LAYER_SIZE = 512
OUTPUT_LAYER_SIZE = 3

# Create the model
model = tf.keras.models.Sequential(
    [
        # First convolutional layer with max pooling
        tf.keras.layers.Conv2D(
            FIRST_LAYER_FILTERS, (3, 3), activation="relu", input_shape=INPUT_SHAPE
        ),
        tf.keras.layers.MaxPooling2D(2, 2),
        # Second convolutional layer with max pooling
        tf.keras.layers.Conv2D(SECOND_LAYER_FILTERS, (3, 3), activation="relu"),
        tf.keras.layers.MaxPooling2D(2, 2),
        # Third convolutional layer with max pooling
        tf.keras.layers.Conv2D(THIRD_LAYER_FILTERS, (3, 3), activation="relu"),
        tf.keras.layers.MaxPooling2D(2, 2),
        # Fourth convolutional layer with max pooling
        tf.keras.layers.Conv2D(FOURTH_LAYER_FILTERS, (3, 3), activation="relu"),
        tf.keras.layers.MaxPooling2D(2, 2),
        # Input layer to convert the 3D feature maps to 1D feature vectors
        tf.keras.layers.Flatten(),
        # Hidden layer
        tf.keras.layers.Dense(DENSE_LAYER_SIZE, activation="relu"),
        # Output layer
        tf.keras.layers.Dense(OUTPUT_LAYER_SIZE, activation="softmax"),
    ]
)

In [13]:
# Check the summary of the model
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 74, 74, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 72, 72, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 36, 36, 64)        0         
 g2D)                                                            
                                                                 
 conv2d_2 (Conv2D)           (None, 34, 34, 128)       73856     
                                                                 
 max_pooling2d_2 (MaxPoolin  (None, 17, 17, 128)       0

## Model Summary Explanation

The model is a Sequential model, which means it's a linear stack of layers. Here's a summary of each layer:

1. **Conv2D Layer**: This is a 2D convolution layer. It has 32 output filters and uses a 3x3 kernel. The output shape is (148, 148, 32). It has 896 parameters.

2. **MaxPooling2D Layer**: This is a max pooling layer for spatial data. It reduces the spatial size (height, width) of the input and helps to prevent overfitting. The output shape is (74, 74, 32).

3. **Conv2D Layer**: Another 2D convolution layer. It has 64 output filters and uses a 3x3 kernel. The output shape is (72, 72, 64). It has 18,496 parameters.

4. **MaxPooling2D Layer**: Another max pooling layer. The output shape is (36, 36, 64).

5. **Conv2D Layer**: Another 2D convolution layer. It has 128 output filters and uses a 3x3 kernel. The output shape is (34, 34, 128). It has 73,856 parameters.

6. **MaxPooling2D Layer**: Another max pooling layer. The output shape is (17, 17, 128).

7. **Conv2D Layer**: Another 2D convolution layer. It has 256 output filters and uses a 3x3 kernel. The output shape is (15, 15, 256). It has 295,168 parameters.

8. **MaxPooling2D Layer**: Another max pooling layer. The output shape is (7, 7, 256).

9. **Flatten Layer**: This layer flattens the input. It does not affect the batch size. The output shape is (12544).

The model has a total of 6,812,995 parameters, all of which are trainable. The size of the model is approximately 25.99 MB.


In [14]:
# Constants for model.compile
OPTIMIZER = tf.optimizers.RMSprop()
LOSS = "kullback_leibler_divergence"
METRICS = ["accuracy"]

# Compile the model
model.compile(optimizer=OPTIMIZER, loss=LOSS, metrics=METRICS)

In [15]:
# Constants for model.fit
EPOCHS = 15

# Train the model
model.fit(
    train_generator, epochs=EPOCHS, validation_data=validation_generator, verbose=2
)

Epoch 1/15


108/108 - 60s - loss: 0.7502 - accuracy: 0.6659 - val_loss: 0.2421 - val_accuracy: 0.9092 - 60s/epoch - 560ms/step
Epoch 2/15
108/108 - 58s - loss: 0.2661 - accuracy: 0.9071 - val_loss: 0.2689 - val_accuracy: 0.8839 - 58s/epoch - 541ms/step
Epoch 3/15
108/108 - 67s - loss: 0.1806 - accuracy: 0.9398 - val_loss: 0.1865 - val_accuracy: 0.9458 - 67s/epoch - 622ms/step
Epoch 4/15


KeyboardInterrupt: 

In [ ]:
# Create a root window and hide it
root = tk.Tk()
root.withdraw()

# Open the file dialog and get the selected file's path
file_path = filedialog.askopenfilename()

# Define a dictionary to map class indices to names
class_dict = {0: "paper", 1: "rock", 2: "scissors"}


def predict_image_class(file_path, model):
    """Predict the class of an image."""
    # Load and resize the image
    img = image.load_img(file_path, target_size=(150, 150))

    # Display the image
    plt.imshow(img)
    plt.show()

    # Convert the image to an array and add an extra dimension
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)

    # Stack the images
    images = np.vstack([x])

    # Predict the class of the image
    classes = model.predict(images, batch_size=10)

    # Find the class with the highest score
    class_index = np.argmax(classes[0])

    # Return the name of the class
    return class_dict[class_index]


# Predict the class of the selected image
class_name = predict_image_class(file_path, model)
print(f"The image {file_path} is predicted to be {class_name}.")

NameError: name 'tk' is not defined